jetty-infinispan/infinispan-common/src/main/java/org/eclipse/jetty/session/infinispan/SessionDataMarshaller.java

```diff
@@ -49,6 +67,33 @@ public String getTypeName()
         return "org_eclipse_jetty_session_infinispan.InfinispanSessionData";
     }
 
+    @Override
+    public InfinispanSessionData readObject(ObjectInput input) throws IOException, ClassNotFoundException
+    {
+        SerializationContext serializationContext = initSerializationContext();
+
+        // invokes readFrom(ProtoStreamReader)
+        InfinispanSessionData data = ProtobufUtil.readFrom(serializationContext, new BoundDelegatingInputStream(input),
+                InfinispanSessionData.class);
+        if (data != null)
+        {
+            data.deserializeAttributes();
+        }
+        return data;
+    }
+
+    @Override
+    public void writeObject(ObjectOutput output, InfinispanSessionData object) throws IOException
+    {
+        SerializationContext serializationContext = initSerializationContext();
+
+     // invokes writeTo(ProtoStreamWriter, InfinispanSessionData)
+        byte[] data = ProtobufUtil.toByteArray(serializationContext, object);
+        int length = data.length;
+        output.write(ByteBuffer.allocate(4).putInt(length).array());
+        output.write(data);        
+    }
+
```

Comment:
Can you explain the necessity for this special write? And also the BoundDelegatingInputStream?

Reply:
I found out that when reading the session from a file, the returned ObjectInput has more data than writeObject has written. It seems Infinispan appends some data at the end for some reason.

If I just wrap all the data in an InputStream, Protobuf fails to deserialize the session. So to know exactly how many bytes need to be parsed, the write first saves the length of the data.

BoundDelegatingInputStream is just a wrapper that first reads the length of the data and stops when the length is reached.

Reply:
So if you configure storeAsBinary, but don't use a filestore, what happens? Is that combination even possible?

Reply:
Sure is possible. storeAsBinary means that the data is stored serialized in the cache (memory). The file is always binary

This means that when you put an instance in the cache the instance is first serialized and than saved. When you get an instance from the cache it has to be deserialized first. In this case the data from the cache is just copied to a file

If storeAsBinary is disabled, you put and get the instance itself in the cache. The data is serialized only when it is moved from the cache to the file

Reply:
If storeAsBinary=true but not filestore, which form of serialization is used?

Reply:
Always the Externalizer interface

Reply:
Store as binary is explained in the user guide:

https://infinispan.org/docs/9.4.x/user_guide/user_guide.html#store_binary

Reply:
I wish infinispan had a unified serialization api!

I think we're good with your solution (just make the serialization context a lazily initialized static).